### PR TITLE
Additional Virtual Staining Metrics

### PR DESCRIPTION
**Pull request by MattsonCam**
This pr integrates additional virtual staining metrics.

### CODE REVIEW

File: metrics/PearsonCorrelation.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        denominator = torch.sqrt(
+            (preds_centered.pow(2).sum(dim=1) * targets_centered.pow(2).sum(dim=1))
+        )
+        per_sample_pearson = torch.where(
+            denominator > 0,
+            numerator / denominator,
+            torch.tensor(0.0, device=self.device),
+        )
+
+        self.total_pearson += per_sample_pearson.sum().detach()
```
Comment (wli51) [inline metrics/PearsonCorrelation.py:77]: Pearson correlation is computed by $r = \frac{\sum_i(x_i-\overline{x})(y_i-\overline{y})}{\sqrt{\sum_i(x_i-\overline{x})^2}\sqrt{\sum_i(y_i-\overline{y})^2}}$

Note that to compute the accurate pearson correlation for a group of image pred+truth pairs in a mini batch setting is not guarantteed to be equivalent to accumulating a weighted average of mini batch pearson correlation as the mini batch $\overline{x}$ $\overline{y}$ are almost surely different from the entire population means of x and y. I think what you are computing here is a mean of pearson correlation estimator as opposed to the real pearson correlation. 

Conceptually for the sake of virtual staining I also am unsure which of the variant of pearson correlation makes the most sense. I am guessing the real pearson correlation aggregated across dataset/subset is what you want to report as scalar to showcase how well performing your model is. Whereas if we want to use pcc for training then your current implementation is probably best for quicker and more stable convergence as the global pcc risks forcing your model to remember the entire dataset which may or may not be good.  

With that said, the standard way to real pearson cc in mini batches involves storing (accumulating) all of the following across batches:
- n
- $\sum x$ # sum of all flattened pred pixels 
- $\sum x^2$ also sum of squared pixel values 
- $\sum y$ # sum of all flattened truth pixels
- $\sum y^2$ # also sum of all squared y pixel values
- $\sum xy$ # per pixel location product of pred and truth pixel values summed and accumulated

and in the end the correlation value can be aggregated as
$r = \frac{\sum xy - \frac{\sum x \sum y}{n}}{ \sqrt{ (\sum x^2 - (\sum x)^2 / n) (\sum y^2 - (\sum y)^2 / n) } }$ which is just an expanded and cleaned up version of the original formula. 

I believe torch metrics also implements their own pcc but i am unsure if it is sample mean person correlation estimator or the "real" one.

**Review by wli51 — APPROVED** [review]
LGTM. Your current PCC implementation may or may not be what you wanted to compute so perhaps it deserves some attention before merging. Please see my comment for details.

File: metrics/PearsonCorrelation.py (hunk 1 above)
Comment (MattsonCam) [reply metrics/PearsonCorrelation.py:77]: To be honest I wasn't sure about this either, but I decided to use the average pearsons correlation across samples because I do the same thing for PSNR and SSIM. I think it is more useful to isolate variation in metrics per nuclei in this case, so I will need to change the L1 and L2 metrics to not be globally computed. Plus so metrics, such as DISTS, can't be computed globally without limitations so I think it makes sense to keep metric computations consistent in this sense. I may compute the standard deviation of these metrics in another pr.

I also tried the torch metrics version of this, but it didn't seem to work for computing per-image correlation after flattening the last three dimensions.

File: metrics/L1.py
Comment (wli51) [inline metrics/L1.py:1]: I think the modification made sense. I just wanted to note that mathematically the per sample L1 mean here would be nearly the same as just calling plain `torch.L1Loss` with `reduction='mean'` and they would only be noticeably different if we have a very small sample size relative to batch size and lacking of data shuffling. So maybe the end result average will not be fundamentally different. I imagine if richness in logging, or interest in applying a focal style loss where the single samples that have higher loss values gets weighted more, are demanded, computing of per sample and or per channel l1 would be more useful?

File: metrics/L2.py
Comment (wli51) [inline metrics/L2.py:1]: the same comments from L1 also applies here

File: metrics/LPIPS.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+from .AbstractMetric import AbstractMetric
+
+
+class LPIPS(AbstractMetric):
+    """LPIPS metric with epoch accumulation support."""
+
+    def __init__(
+        self,
+        net: str = "vgg",
```
Comment (wli51) [inline metrics/LPIPS.py]: maybe this is best made a literal?

File: metrics/LPIPS.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
+from typing import Union
+
+import lpips
```
Comment (wli51) [inline metrics/LPIPS.py]: would recommend the torchmetrics `LearnedPerceptualImagePatchSimilarity` over the author implementation as the torchmetrics is more well maintained, has more similar signature as regular torch loss, has built in state like any loss object should. I think the only advantage of the original implementation is that it better exposes the intermediate compute feature maps and tuning of things like normalization.

**Review by wli51 — APPROVED** [review]
LGTM

File: metrics/L1.py
Comment (MattsonCam) [reply metrics/L1.py:1]: I think that's right. I use `reduction='mean'` when I compute L1 as a loss for back-propagation, so if I decide to weight samples then I'll probably need to update this part. The arithmetic mean of the image metric values gives each sample an equal metric weight contribution, and computing them individually allows me to calculate the standard deviation of the metric values as well. At one point I was think of logging metrics for different strata, but we are trying to start training models , so that would add more complexity and time to the implementation. Plus, I can always use the saved model for these epochs to assess performance after training. These datasets contain different cell types and treatments, so this would add another layer of difficulty.

File: metrics/LPIPS.py (hunk 2 above)
Comment (MattsonCam) [reply metrics/LPIPS.py]: Good suggestion, I will implement this

File: metrics/LPIPS.py (hunk 3 above)
Comment (MattsonCam) [reply metrics/LPIPS.py]: Good idea, I will implement this

**Comment by MattsonCam** [comment]
Thanks @wli51 for the reviews! Merging now
